### DISCUSSION TITLE
Understanding the 'heads' attribute in a Doc type

### DISCUSSION THREAD

**Discussion by JeremyCraigMartinez** [explosion/spaCy#12578] [Help: Other Questions] [answered]
The [documentation](https://spacy.io/api/doc) states the following:

> A list of values, of the same length as words, to assign as the head for each word. Head indices are the absolute position of the head in the Doc. Defaults to None.

But I placed a `print(...)` in an augmenter to inspect what the `heads` are and I can't make sense of it

Example sentence: **This is an example sentence of random length. One more sentence of a different length, with a tad more punctuation!!**

Here is a list of each individual word with their corresponding `HEAD` value:

```
This 1
is 1
an 4
example 4
sentence 1
of 4
random 7
length 5
. 1
One 11
more 11
sentence 11
of 11
a 15
different 15
length 12
, 11
with 11
a 19
tad 20
more 21
punctuation 17
! 11
! 11
```

Maybe just a rephrasing would help, because I can't quite decipher what this is telling me 😅

**Comment by adrianeboyd** [comment, marked as answer]
The head is the same as if you'd looked up that index in the doc with `doc[head.i]`, so the head of "This" is "is", "is" is the head of itself (ROOT), the head of "an" is "sentence", etc.

**Reply by JeremyCraigMartinez** [reply]
I'm still grasping parts of speech. And while I'm gonna mark this as an answer, I came cross an answer in the chapter 1 of the [Advanced NLP with spaCy course](https://spacy.io/usage/spacy-101/) that gave me the lightbulb moment:

> "The `.head` attribute returns the syntactic head token. You can also think of it as the parent token this word is attached to."

```python
>>> for token in doc:
...     print(token.text, token.pos_, token.dep_, token.head.text)
... 
She PRON nsubj ate
ate VERB ROOT ate
the DET det pizza
pizza NOUN dobj ate
```

**Reply by adrianeboyd** [reply]
We can also recommend this online textbook as a good introduction for a lot of NLP topics: https://web.stanford.edu/~jurafsky/slp3/
